google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudLibraryProjectState.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.apis;
+
+import com.google.cloud.tools.libraries.json.CloudLibrary;
+import com.google.common.collect.ImmutableSet;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.components.ProjectComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Computable;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.fest.util.Maps;
+import org.fest.util.VisibleForTesting;
+import org.jetbrains.idea.maven.dom.MavenDomUtil;
+import org.jetbrains.idea.maven.dom.model.MavenDomDependency;
+import org.jetbrains.idea.maven.dom.model.MavenDomProjectModel;
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+import org.jetbrains.idea.maven.project.MavenProjectsManager.Listener;
+
+/**
+ * A {@link ProjectComponent} that maintains, in memory, the current set of {@link CloudLibrary}
+ * that are configured on the user's project.
+ */
+public class CloudLibraryProjectState implements ProjectComponent {
+
+  private static final Logger logger = Logger.getInstance(AddCloudLibrariesAction.class);
+
+  private final Project project;
+  private Map<Module, Set<CloudLibrary>> moduleLibraryMap = Maps.newHashMap();
+  private List<CloudLibrary> allLibraries;
+
+  private CloudLibraryProjectState(Project project) {
```

Comment:
Interesting, so the IDE is able to instantiate this class from private constructor?

Reply:
yeah, I thought it was odd too (I initially had it public). But whatever they are doing allows for it to be private.